File: -51-N_Queens/Code.py
class Solution:
    def isSafe(self, board, row, col):
        x, y = row, col

        # Check left side of the current row
        while y >= 0:
            if board[x][y] == 1:
                return False
            y -= 1

        # Check upper diagonal on left side
        x, y = row, col
        while x < len(board) and y >= 0:
            if board[x][y] == 1:
                return False
            x += 1
            y -= 1

        # Check lower diagonal on left side
        x, y = row, col
        while x >= 0 and y >= 0:
            if board[x][y] == 1:
                return False
            x -= 1
            y -= 1

        return True

    def addSol(self, board, n, ans):
        arr = []
        for i in range(n):
            row_str = ''.join('Q' if board[i][j] == 1 else '.' for j in range(n))
            arr.append(row_str)
        ans.append(arr)

    def solve(self, board, n, col, v, ans):
        if col == n:
            self.addSol(board, n, ans)
            return

        for row in range(n):
            if self.isSafe(board, row, col):
                board[row][col] = 1
                self.solve(board, n, col + 1, v, ans)
                board[row][col] = 0

    def solveNQueens(self, n):
        board = [[0] * n for _ in range(n)]
        v = []
        ans = []
        self.solve(board, n, 0, v, ans)
        return ans

File: -51-N_Queens/Code.cpp
class Solution {
public:
    bool isSafe(vector<vector<int>>& board, int row, int col) {
        int x = row, y = col;
        while(y >= 0) {
            if(board[x][y--] == 1)  return false;
        }
        y = col;
        while(x < board.size() && y >= 0) {
            if(board[x++][y--] == 1)    return false;
        }
        x = row, y = col;
        while(x >= 0 && y >= 0) {
            if(board[x--][y--] == 1)    return false;
        }
        return true;
    }

    void addSol(vector<vector<int>>& board, int n, vector<vector<string>>& ans) {
        vector<string>arr;
        for(int i = 0; i < n; i++) {
            string str = "";
            for(int j = 0;  j< n; j++) {
                if(board[i][j] == 1)        str.push_back('Q');
                else    str.push_back('.');
            }
            arr.push_back(str);
        }
        ans.push_back(arr);
    }

    void solve(vector<vector<int>>& board, int n, int col, vector<string>& v, vector<vector<string>>& ans) {
        if(col == n) {
            addSol(board, n, ans);
            return ;
        }
        for(int row = 0; row < n; row++) {
            if(isSafe(board, row, col)) {
                board[row][col] = 1;
                solve(board, n, col + 1, v, ans);
                board[row][col] = 0;
            }
        }
    }

    vector<vector<string>> solveNQueens(int n) {
        vector<vector<int>>board(n, vector<int>(n, 0));
        vector<string>v;
        vector<vector<string>>ans;
        solve(board, n, 0, v, ans);
        return ans;
    }
};
